src/tune-hyperparameters/src/main/scala/HyperparamBuilder.scala

```diff
@@ -0,0 +1,81 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import org.apache.spark.ml.param._
+
+import scala.collection.mutable
+import scala.util.Random
+
+abstract class HyperParam[T] {
+  def getNextValue(): T
+  def getParamPair(param: Param[_]): ParamPair[T] = {
+    ParamPair(param.asInstanceOf[Param[T]], getNextValue())
+  }
+}
+
+abstract class RangeHyperParam[T](val min: T, val max: T, val seed: Long) extends HyperParam[T] {
+  val random = new Random(seed)
+}
+
+class IntRangeHyperParam(min: Int, max: Int, seed: Long = 0)
+  extends RangeHyperParam[Int](min, max, seed) {
+  def getNextValue(): Int = {
+    val range = max - min
+    random.nextInt(range) + min
+  }
+}
+
+class LongRangeHyperParam(min: Long, max: Long, seed: Long = 0)
+  extends RangeHyperParam[Long](min, max, seed) {
+  def getNextValue(): Long = {
+    val range = max - min
+    (random.nextLong() * range) + min
+  }
+}
+
+class FloatRangeHyperParam(min: Float, max: Float, seed: Long = 0)
+  extends RangeHyperParam[Float](min, max, seed) {
+  val doubleRange = new DoubleRangeHyperParam(min.toDouble, max.toDouble)
+  def getNextValue(): Float = {
+    doubleRange.getNextValue().toFloat
+  }
+}
+
+class DoubleRangeHyperParam(min: Double, max: Double, seed: Long = 0)
+  extends RangeHyperParam[Double](min, max, seed) {
+  def getNextValue(): Double = {
+    val range = max - min
+    (random.nextDouble() * range) + min
+  }
+}
+
+class DiscreteHyperParam[T](values: Array[T], seed: Long = 0) extends HyperParam[T] {
+  val random = new Random(seed)
+
+  def getNextValue(): T = {
+    values(random.nextInt(values.length))
+  }
+}
+
+/** Specifies the search space for hyperparameters.
+  */
+class HyperparamBuilder {
+
+  private val hyperparams = mutable.Map.empty[Param[_], HyperParam[_]]
+
+  /**
+    * Adds a param to the search space.
+    */
+  def addHyperparam[T](param: Param[T], values: HyperParam[T]): this.type = {
+    hyperparams.put(param, values)
+    this
+  }
+
+  /**
+    * Builds the search space of hyperparameters.
+    * @return The map of hyperparameters to search through.
```

Comment:
This is a good start, we should think about this problem as sampling from distributions over different domains. In this light these would all be uniform distributions see:

http://scikit-learn.org/stable/modules/grid_search.html#randomized-parameter-search

For an example of a good API

Reply:
Yes!  I want to add more distributions in the future.  It can be put on the hyperparams class constructor.